Homework 1 of ML Zoomcamp
Created by Radenpm9

Q1. Pandas Version

Checking my pandas version

In [49]:
import pandas as pd

In [50]:
pd.__version__

'2.3.2'

My Pandas is version 2.3.2

Now reading the data after downloading it

In [51]:
#read the data
df = pd.read_csv('car_fuel_efficiency.csv')
df

engine_displacement  num_cylinders  horsepower  vehicle_weight  \
0                     170            3.0       159.0     3413.433759   
1                     130            5.0        97.0     3149.664934   
2                     170            NaN        78.0     3079.038997   
3                     220            4.0         NaN     2542.392402   
4                     210            1.0       140.0     3460.870990   
...                   ...            ...         ...             ...   
9699                  140            5.0       164.0     2981.107371   
9700                  180            NaN       154.0     2439.525729   
9701                  220            2.0       138.0     2583.471318   
9702                  230            4.0       177.0     2905.527390   
9703                  270            3.0       140.0     2908.043477   

      acceleration  model_year  origin fuel_type         drivetrain  \
0             17.7        2003  Europe  Gasoline    All-wheel drive   
1             17.8        2007     USA  Gasoline  Front-wheel drive   
2             15.1        2018  Europe  Gasoline  Front-wheel drive   
3             20.2        2009     USA    Diesel    All-wheel drive   
4             14.4        2009  Europe  Gasoline    All-wheel drive   
...            ...         ...     ...       ...                ...   
9699          17.3        2013  Europe    Diesel  Front-wheel drive   
9700          15.0        2004     USA  Gasoline    All-wheel drive   
9701          15.1        2008     USA    Diesel    All-wheel drive   
9702          19.4        2011     USA    Diesel  Front-wheel drive   
9703          14.7        2005    Asia    Diesel    All-wheel drive   

      num_doors  fuel_efficiency_mpg  
0           0.0            13.231729  
1           0.0            13.688217  
2           0.0            14.246341  
3           2.0            16.912736  
4           2.0            12.488369  
...         ...                  ...  
9699        NaN            15.101802  
9700        0.0            17.962326  
9701       -1.0            17.186587  
9702        1.0            15.331551  
9703       -1.0            14.884467  

[9704 rows x 11 columns]

Q2 Record counts  
How many records are in the dataset?  
It is mentioned that the data has 9704 rows and 11 columns.  
So the records in the data set are 9704.

Q3 Fuel types  
How many fuel types are presented in the dataset?  

In [52]:
#checking the unique number in all columns
df.nunique()

engine_displacement      36
num_cylinders            14
horsepower              192
vehicle_weight         9704
acceleration            162
model_year               24
origin                    3
fuel_type                 2
drivetrain                2
num_doors                 9
fuel_efficiency_mpg    9704
dtype: int64

We can see that the fuel_type has 2 type of fuel.

Q4 Missing Values  
How many columns in the dataset have missing values?

In [53]:
#checking the missing values
df.isnull().sum()

engine_displacement      0
num_cylinders          482
horsepower             708
vehicle_weight           0
acceleration           930
model_year               0
origin                   0
fuel_type                0
drivetrain               0
num_doors              502
fuel_efficiency_mpg      0
dtype: int64

We can see that we have missing values in 4 columns: num_cylinders, horsepower, acceleration, num_doors

Q5. Max fuel efficiency  
What's the maximum fuel efficiency of cars from Asia?


In [54]:
#finding out cars from Asia
car_asia = df[df['origin'] == 'Asia']
car_asia

engine_displacement  num_cylinders  horsepower  vehicle_weight  \
8                     250            1.0       174.0     2714.219310   
12                    320            5.0       145.0     2783.868974   
14                    200            6.0       160.0     3582.687368   
20                    150            3.0       197.0     2231.808142   
21                    160            4.0       133.0     2659.431451   
...                   ...            ...         ...             ...   
9688                  260            4.0         NaN     3948.404625   
9692                  180            3.0       188.0     3680.341381   
9693                  280            2.0       148.0     2545.070139   
9698                  180            1.0       131.0     3107.427820   
9703                  270            3.0       140.0     2908.043477   

      acceleration  model_year origin fuel_type         drivetrain  num_doors  \
8             10.3        2016   Asia    Diesel  Front-wheel drive       -1.0   
12            15.1        2010   Asia    Diesel    All-wheel drive        1.0   
14            14.9        2007   Asia    Diesel    All-wheel drive        0.0   
20            18.7        2011   Asia  Gasoline  Front-wheel drive        1.0   
21             NaN        2016   Asia  Gasoline  Front-wheel drive       -1.0   
...            ...         ...    ...       ...                ...        ...   
9688          15.5        2018   Asia    Diesel    All-wheel drive       -1.0   
9692          18.0        2016   Asia  Gasoline  Front-wheel drive        1.0   
9693          15.6        2012   Asia    Diesel    All-wheel drive        0.0   
9698          13.2        2005   Asia  Gasoline  Front-wheel drive       -2.0   
9703          14.7        2005   Asia    Diesel    All-wheel drive       -1.0   

      fuel_efficiency_mpg  
8               16.823554  
12              16.175820  
14              11.871091  
20              18.889083  
21              16.077730  
...                   ...  
9688            11.054830  
9692            11.711653  
9693            17.202782  
9698            13.933716  
9703            14.884467  

[3247 rows x 11 columns]

In [55]:
max_car_asia = car_asia['fuel_efficiency_mpg'].max()
max_car_asia

np.float64(23.759122836520497)

So, the max efficiency of Asian car is 23.75

Q6. Median value of horsepower

    1. Find the median value of horsepower column in the dataset.
    2. Next, calculate the most frequent value of the same horsepower column.
    3. Use fillna method to fill the missing values in horsepower column with the most frequent value from the previous step.
    4. Now, calculate the median value of horsepower once again.

Has it changed?

In [56]:
#finding median value of horsepower
median_hp = df['horsepower'].median()
median_hp

np.float64(149.0)

In [57]:
#finding the most frequent value (mode) of horsepower
modes_hp = df['horsepower'].mode()
modes_hp

0    152.0
Name: horsepower, dtype: float64

In [58]:
#replace missing value with modes
hp_new = df['horsepower'].fillna(152)
#checking median value of newly replaced missing value
median_hp_new = hp_new.median()
median_hp_new

np.float64(152.0)

In [59]:
#checking the descriptive statistics of old dataset before .fillna()
df['horsepower'].describe()

count    8996.000000
mean      149.657292
std        29.879555
min        37.000000
25%       130.000000
50%       149.000000
75%       170.000000
max       271.000000
Name: horsepower, dtype: float64

The 50% mean median = 149. Same as result from .median() 

In [60]:
#checking the descriptive statistics of new dataset after .fillna()
hp_new.describe()

count    9704.000000
mean      149.828215
std        28.775248
min        37.000000
25%       131.000000
50%       152.000000
75%       168.000000
max       271.000000
Name: horsepower, dtype: float64

the 50% value is 152. this confirm the previous result.  
So, Te median value is increased.

Q7. Sum of weights

    1. Select all the cars from Asia
    2. Select only columns vehicle_weight and model_year
    3. Select the first 7 values
    4. Get the underlying NumPy array. Let's call it X.
    5. Compute matrix-matrix multiplication between the transpose of X and X. To get the transpose, use X.T. Let's call the result XTX.
    6. Invert XTX.
    7. Create an array y with values [1100, 1300, 800, 900, 1000, 1100, 1200].
    8. Multiply the inverse of XTX with the transpose of X, and then multiply the result by y. Call the result w.
    9. What's the sum of all the elements of the result?


In [61]:
#select all the cars from Asia. I have done it, check above
car_asia

engine_displacement  num_cylinders  horsepower  vehicle_weight  \
8                     250            1.0       174.0     2714.219310   
12                    320            5.0       145.0     2783.868974   
14                    200            6.0       160.0     3582.687368   
20                    150            3.0       197.0     2231.808142   
21                    160            4.0       133.0     2659.431451   
...                   ...            ...         ...             ...   
9688                  260            4.0         NaN     3948.404625   
9692                  180            3.0       188.0     3680.341381   
9693                  280            2.0       148.0     2545.070139   
9698                  180            1.0       131.0     3107.427820   
9703                  270            3.0       140.0     2908.043477   

      acceleration  model_year origin fuel_type         drivetrain  num_doors  \
8             10.3        2016   Asia    Diesel  Front-wheel drive       -1.0   
12            15.1        2010   Asia    Diesel    All-wheel drive        1.0   
14            14.9        2007   Asia    Diesel    All-wheel drive        0.0   
20            18.7        2011   Asia  Gasoline  Front-wheel drive        1.0   
21             NaN        2016   Asia  Gasoline  Front-wheel drive       -1.0   
...            ...         ...    ...       ...                ...        ...   
9688          15.5        2018   Asia    Diesel    All-wheel drive       -1.0   
9692          18.0        2016   Asia  Gasoline  Front-wheel drive        1.0   
9693          15.6        2012   Asia    Diesel    All-wheel drive        0.0   
9698          13.2        2005   Asia  Gasoline  Front-wheel drive       -2.0   
9703          14.7        2005   Asia    Diesel    All-wheel drive       -1.0   

      fuel_efficiency_mpg  
8               16.823554  
12              16.175820  
14              11.871091  
20              18.889083  
21              16.077730  
...                   ...  
9688            11.054830  
9692            11.711653  
9693            17.202782  
9698            13.933716  
9703            14.884467  

[3247 rows x 11 columns]

In [62]:
#Select only columns vehicle_weight and model_year
car_asia[['vehicle_weight', 'model_year']]

vehicle_weight  model_year
8        2714.219310        2016
12       2783.868974        2010
14       3582.687368        2007
20       2231.808142        2011
21       2659.431451        2016
...              ...         ...
9688     3948.404625        2018
9692     3680.341381        2016
9693     2545.070139        2012
9698     3107.427820        2005
9703     2908.043477        2005

[3247 rows x 2 columns]

In [63]:
#Select the first 7 values
car_asia[['vehicle_weight', 'model_year']].head(7)

vehicle_weight  model_year
8      2714.219310        2016
12     2783.868974        2010
14     3582.687368        2007
20     2231.808142        2011
21     2659.431451        2016
34     2844.227534        2014
38     3761.994038        2019

In [64]:
#Get the underlying NumPy array. Let's call it X.
X = car_asia[['vehicle_weight', 'model_year']].head(7).values
X

array([[2714.21930965, 2016.        ],
       [2783.86897424, 2010.        ],
       [3582.68736772, 2007.        ],
       [2231.8081416 , 2011.        ],
       [2659.43145076, 2016.        ],
       [2844.22753389, 2014.        ],
       [3761.99403819, 2019.        ]])

In [70]:
#Compute matrix-matrix multiplication between the transpose of X and X. To get the transpose, use X.T. Let's call the result XTX.
XT = X.T
XT

array([[2714.21930965, 2783.86897424, 3582.68736772, 2231.8081416 ,
        2659.43145076, 2844.22753389, 3761.99403819],
       [2016.        , 2010.        , 2007.        , 2011.        ,
        2016.        , 2014.        , 2019.        ]])

In [71]:
#the multiplication
XTX = np.dot(XT, X)
XTX

array([[62248334.33150762, 41431216.50732678],
       [41431216.50732678, 28373339.        ]])

In [72]:
import numpy as np
from numpy.linalg import inv
#Invert XTX.
inv_XTX = inv(XTX)
inv_XTX

array([[ 5.71497081e-07, -8.34509443e-07],
       [-8.34509443e-07,  1.25380877e-06]])

In [33]:
#Create an array y with values [1100, 1300, 800, 900, 1000, 1100, 1200].
y = np.array([1100, 1300, 800, 900, 1000, 1100, 1200])
y

array([1100, 1300,  800,  900, 1000, 1100, 1200])

In [74]:
#Multiply the inverse of XTX with the transpose of X, and then multiply the result by y. Call the result w.
w = np.dot(np.dot(inv_XTX, XT),y)
w

array([0.01386421, 0.5049067 ])

In [75]:
#What's the sum of all the elements of the result?
w.sum()

np.float64(0.5187709081074001)

So, the sum of all the elements of the result is 0.518